# Question 2: International Coding Competition

Name this Jupyter Notebook as: <br>
`TASK2_<your name>_<class>_<index number>.ipynb`

For each of the sub-tasks, add a comment statement at the beginning of the code using the hash symbol '#' to indicate the sub-task the program code belongs to.

## Task 2.1
Write an insertion sort function that takes three parameters: `data_list`,`index`,`ascending` and returns a sorted `data_list`.

* `data_list` is a list of tuples.
* `index` is an integer to specify an index within a tuple.
* `ascending` is a Boolean flag.

The function will sort `data_list` base on the values at the specified `index` within the tuples. The `ascending` parameter determins the order of the sort: `True` for ascending order, `False` for descending order.

The function specification is:

```python
FUNCTION insertion_sort(data_list: List[Tuple], index: Integer, ascending:Boolean) RETURN List[Tuple]
```
| Note: The syntax that follows the positional argument is just type hinting.

<div style="text-align: right">[6]</div>

> Example of `insertion_sort` function call:
>```python
> data_list = [(1,'Alice', 85), (2, 'Bob',90), (3,'Charlie',80)]
> sorted_list = insertion_sort(data_list, 2, ascending=False)
> print(sorted_list)
>
> # Output: [(2,'Bob',90), (1,'Alice',85), (3,'Charlie',80)]
>```



In [21]:
# Task 2.1      
def insertion_sort(data_list: list[tuple], index: int, ascending:bool=True) -> list[tuple]:
    sorted_list = []
    for data in data_list:
        i = 0
        try: 
            while bool(data[index] < sorted_list[i][index]) != bool(ascending):
                i += 1
        except IndexError: pass
        sorted_list.insert(i, data)
    return sorted_list

data_list = [(1,'Alice', 85), (2, 'Bob',90), (3,'Charlie',80)]
sorted_list = insertion_sort(data_list, 2, ascending=False)
print(sorted_list)

[(2, 'Bob', 90), (1, 'Alice', 85), (3, 'Charlie', 80)]


## Task 2.2
At an international coding competition, the results of the participants were recorded in a text file `CODING_COMPETITION.txt`.

Each record int he file has the following format, with each field separated by a comma:
`<participant_id>,<participant_name>,<country>,<score>`

Write a function `task2_2` that takes a `filename` as parameter, reads the comma-separated data from the file and returns the data as a list of tuples:

> Example: After calling function `task2_2`, these sample records in the file: <br>
>
> `101,Jong Tan,Singapore,210` <br>
> `102,Mei Ling,Singapore,190` <br>
>
> will be returned as <br>
> `[('101','Jong Tan','Singapore',210),('102','Mei Ling','Singapore',190)]`

Test your function by calling `task2_2("CODING_COMPETITION.txt")`
<div style="text-align: right">[5]<div>

In [10]:
# Task 2.2
import csv

def task2_2(filename: str):
    with open('./Resources/TASK2/'+filename) as f:
        return list(map(tuple, list(csv.reader(f))))

task2_2("CODING_COMPETITION.txt")


[('101', 'Jong Tan', 'Singapore', '210'),
 ('102', 'Mei Ling', 'Singapore', '190'),
 ('103', 'John Smith', 'USA', '230'),
 ('104', 'Emily Davis', 'USA', '85'),
 ('105', 'Chen Wei', 'China', '220'),
 ('106', 'Li Na', 'China', '105'),
 ('107', 'Kumar Patel', 'India', '100'),
 ('108', 'Arjun Rao', 'India', '95'),
 ('109', 'Alice Johnson', 'UK', '240'),
 ('110', 'Robert Brown', 'UK', '70'),
 ('111', 'Marco Rossi', 'Italy', '225'),
 ('112', 'Giulia Ferrari', 'Italy', '60'),
 ('113', 'Ahmed Khan', 'Pakistan', '210'),
 ('114', 'Ali Ahmed', 'Pakistan', '80'),
 ('115', 'Olga Ivanova', 'Russia', '220'),
 ('116', 'Dmitry Petrov', 'Russia', '90'),
 ('117', 'Paul Walker', 'Australia', '215'),
 ('118', 'Emily Brown', 'Australia', '85'),
 ('119', 'Julia Schmidt', 'Germany', '235'),
 ('120', 'Matthias Mueller', 'Germany', '55'),
 ('121', 'Yuki Tanaka', 'Japan', '225'),
 ('122', 'Haruka Sato', 'Japan', '95'),
 ('123', 'Karen Wilson', 'Canada', '250'),
 ('124', 'David Wilson', 'Canada', '70'),
 ('125', 

## Task 2.3
Write a function `task2_3` that performs the following:
* Use the list of tuples returned from the function `task2_2`.
* Sort the list by country in alphabetical order, using the `insertion_sort` function
* Within each country, sort the participants by their scores in descending order using the `insertion_sort` function.
* Write the sorted data to a text file `SCORE_BY_COUNTRY.txt` in the format: <br>
`<country>,<participant id>,<participant name>,<score>`.
* Each record should be on a new line

The output file should contain lines of record that is first sorted by country then by score. Sample lines in output file:

<center>

<table border="1">
  <tr style="font-family: 'Courier New', monospace;">
    <td>Australia,249,James Scott, 235<br>
        Australia,267,Liam Taylor, 230<br>
        Australia,139,Daniel Miller, 225<br>
        ...<br>
        Singapore,101,Jong Tan,210<br>
        ...</td>
  </tr>
</table>

</center>

Run your function `task2_3` 
<div style="text-align:right;">[9]</div>

In [ ]:
# Task 2.3

def task2_3():
    data = task2_2("CODING_COMPETITION.txt")
    country_sorted = insertion_sort(data, 2)
    country_set = []
    for row in country_sorted:
        if row[2] in country_set: continue
        country_set.append(row[2])
    country_sorted = map(lambda a: (int(a[0]), a[1], a[2], int(a[3])), country_sorted)
    final_sorted = []
    for country in country_set:
        final_sorted.extend(insertion_sort([row for row in data if row[2] == country], 3, ascending=False))
    for i, row in enumerate(final_sorted):
        try:
            final_sorted[i] = row[2], row[0], row[1], row[3]
        except:
            print(row)
    return list(final_sorted)

task2_3()


[('Australia', '250', 'Emily Johnson', '90'),
 ('Australia', '232', 'Jack Wilson', '90'),
 ('Australia', '178', 'Isabella Brown', '90'),
 ('Australia', '162', 'Chris Brown', '90'),
 ('Australia', '140', 'Laura Smith', '85'),
 ('Australia', '118', 'Emily Brown', '85'),
 ('Australia', '249', 'James Scott', '235'),
 ('Australia', '195', 'James Brown', '235'),
 ('Australia', '177', 'William Harris', '235'),
 ('Australia', '267', 'Liam Taylor', '230'),
 ('Australia', '213', 'Ethan Green', '225'),
 ('Australia', '161', 'Anna Wilson', '225'),
 ('Australia', '139', 'Daniel Miller', '225'),
 ('Australia', '231', 'Charlotte Harris', '220'),
 ('Australia', '117', 'Paul Walker', '215'),
 ('Australia', '196', 'Olivia Taylor', '190'),
 ('Australia', '214', 'Mia Roberts', '175'),
 ('Canada', '146', 'Matthew Johnson', '70'),
 ('Canada', '124', 'David Wilson', '70'),
 ('Canada', '252', 'Michael Brown', '65'),
 ('Canada', '180', 'Joshua Clark', '65'),
 ('Canada', '198', 'John Lee', '55'),
 ('Canada', '2